File: Java/142. Linked List Cycle II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head==null || head.next==null) return null;
        ListNode fast = head.next.next;
        ListNode slow = head.next;
        while(fast!=slow && fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        if(fast == null || fast.next == null) return null;
        ListNode temp=head;
        while(temp!=fast){
            temp=temp.next;
            fast=fast.next;
        }
        return fast;
    }
}

File: Python/142. Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None or head.next==None:
            return None

        fast = head.next.next
        slow = head.next
        
        while fast!=slow and fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next

        if fast == None or fast.next == None:
            return None

        temp=head
        while(temp!=fast):
            temp=temp.next
            fast=fast.next

        return fast
